one-ui/src/main/java/com/marklogic/hub/oneui/controllers/EnvironmentController.java

```diff
@@ -171,4 +117,43 @@ public JsonNode reset() {
     public JsonNode getProjectInfo() {
         return environmentService.getProjectInfo();
     }
+
+    @RequestMapping(value = "/project-upload", method = RequestMethod.POST)
+    @ResponseBody
+    public void uploadProject(@RequestParam("zipfile") MultipartFile uploadedFile) throws Exception {
+        UIDeployListener listener = new UIDeployListener(template, false);
+
+        HubProject project = hubConfig.getHubProject();
+        DataHubProjectUtils.replaceProject(project, uploadedFile.getInputStream(), listener);
+
+        dataHubService.unInstall(new UIDeployListener(template,true));
+        install(project.getProjectDirString(), listener);
+    }
+
+    private void install(String directory, UIDeployListener listener) throws Exception {
+        try {
+            if (StringUtils.isEmpty(directory)) {
+                throw new BadRequestException("Property 'directory', identifying project location, not specified");
+            }
+            // setting the project directory will resolve any relative paths
+            Path directoryPath = Paths.get(directory);
+            if (!directoryPath.isAbsolute()) {
+                throw new ProjectDirectoryException("The Project Directory field requires an absolute path. You entered: " + directory, "Enter the absolute path to an existing local directory.");
```

Comment:
Is this ever going to happen, and what would the user be able to do about it? The user isn't able to enter an absolute path - the user is just uploading a file.

If "directory" is invalid, then the system is misconfigured, and I don't think any user can do anything about that, right? An IT person would need to fix the path for the HubProject and then restart the middle tier.

@ryanjdew @wooldridge Am I correct about this? My impression is that a user doesn't have any control over the project directory, except when the application is first initialized. 

I think that means that the value of "directory" can be assumed to be valid. If it weren't valid, the user would never have gotten to this point, and there's not anything a user can do about it either. I think that means we just need a single try/catch around the this code, which is already in place.

Reply:
@rjrudin i am not sure i understand your input here...please clarify.

Reply:
My impression is that the project directory is not specified by a user, but rather when One UI is first initialized. Thus, if there's an error with the value of the directory, there's nothing that a user can do about it. The only thing to do would be to tell the user that the project directory is not valid and to contact a systems administrator.

Reply:
@rjrudin The directory is specified when the install is called via the install page at /install, which the user is guided to do if a project hasn't previously been installed.

Reply:
Okay - I'm thinking of the use case where someone uploads a project - could be Ernie or Wei Lin - and that user doesn't have the ability to specify the project directory.

Reply:
I'll stand down on this for now though - I think it's something that'll come up in testing though - pinging @bsrikan  for awareness.

Reply:
What @rjrudin says makes sense to me. At the time a project is being uploaded, user is logged in and DH is installed viz. project Directory is valid.
Say DH is installed and project dir is "/Users/userA/myProject".
Upload is basically an option to reinstall, so when a user clicks on Upload button, presumably there is a Finder window to select project zip to be uploaded. This is another place where path is required to the project zip being uploaded. 
@hao1st can you pls clarify that the project dir in question is not for the zip being uploaded. If yes, then project dir is already established and the if condition is never satisfied IMO.